WordPress/WordPressTest/PostRepositoryTests.swift

```diff
@@ -241,6 +241,65 @@ class PostRepositoryTests: CoreDataTestCase {
         }
     }
 
+    func testFetchAllPagesStopsOnEmptyAPIResponse() async throws {
+        // Given two pages of API result: first page returns 100 page instances, and the second page returns an empty result.
+        remoteMock.remotePostsToReturnOnSyncPostsOfType = [
+            try (1...100).map {
+                let post = try XCTUnwrap(RemotePost(siteID: NSNumber(value: $0), status: "publish", title: "Post: Test", content: "This is a test post"))
+                post.type = "page"
+                return post
+            },
+            []
+        ]
+
+        let pages = try await repository.fetchAllPages(statuses: [.publish], in: blogID).value
+        XCTAssertEqual(pages.count, 100)
+    }
+
+    func testFetchAllPagesStopsOnNonFullPageAPIResponse() async throws {
+        // Given two pages of API result: first page returns 100 page instances, and the second page returns 10 (any amount that's less than 100) page instances.
+        remoteMock.remotePostsToReturnOnSyncPostsOfType = [
+            try (1...100).map {
+                let post = try XCTUnwrap(RemotePost(siteID: NSNumber(value: $0), status: "publish", title: "Post: Test", content: "This is a test post"))
+                post.type = "page"
+                return post
+            },
+            try (1...10).map {
+                let post = try XCTUnwrap(RemotePost(siteID: NSNumber(value: $0), status: "publish", title: "Post: Test", content: "This is a test post"))
+                post.type = "page"
+                return post
+            },
+        ]
+
+        let pages = try await repository.fetchAllPages(statuses: [.publish], in: blogID).value
+        XCTAssertEqual(pages.count, 110)
+    }
```

Comment:
Have you considered a test to explicitly check that no more requests are made after an empty or < page size response has been received?

🤔 I guess the fact that `await` triggers sort of guarantees this, but I don't know... I feel like it would be "nice" to test it explicitly. 

_Not a must_. If I had some time, I'd see if I could change the source code to make these tests pass with more requests, but I don't think I'll get to it any time soon.

Reply:
> explicitly check that no more requests are made after an empty or < page size response has been received?

I think that's already covered here. If there are more requests fired, then the fake `PostServiceRemote` API response would make sure either

- an extra 100 or 10 page instances are returned to the `PostRepository` fucntion. Or,
- a failure is returned.

That means the final total pages should be more than 110 and breaks the assertion or throw an error. Given the test passes, we should(?) be fairly certain that there no unnecessary requests are fired.

I guess that's not entirely true. Because if the `fetchAllPages` is implemented in a way that removes duplication or swallows error, then the test would still pass when there _are_ unnecessary requests.

I guess we could to add a `XCTFail` to the error case to be really really sure.

https://github.com/wordpress-mobile/WordPress-iOS/blob/5339b86f6ae66cc608ff3a466c8a6b53a0dbd487/WordPress/WordPressTest/PostRepositoryTests.swift#L350-L354

Reply:
I have added another test in https://github.com/wordpress-mobile/WordPress-iOS/pull/21949/commits/ed781a5a1bb52c36bf0eeeceb040a719c05b2ae1 to ensure when there is an HTTP API error, the fetchAllPages should throw, which means the error is not swallowed and at the same time ensure no extra requests are fired after an empty or < page size response.